diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index a92b747..7517752 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -35,6 +35,7 @@
 #include "wpa_auth.h"
 #include "wps_hostapd.h"
 #include "dpp_hostapd.h"
+#include "common/dpp.h"
 #include "gas_query_ap.h"
 #include "hw_features.h"
 #include "wpa_auth_glue.h"
@@ -392,6 +393,9 @@ static void hostapd_free_hapd_data(struct hostapd_data *hapd)
 #ifdef CONFIG_DPP
 	hostapd_dpp_deinit(hapd);
 	gas_query_ap_deinit(hapd->gas);
+#ifdef CONFIG_DPP2
+	dpp_controller_stop(hapd->iface->interfaces->dpp);
+#endif /* CONFIG_DPP2 */
 #endif /* CONFIG_DPP */
 
 	authsrv_deinit(hapd);
